File: strategy/marvin/agent/defender/manmark.lua
local Base = require "agent/base/behavior"
local ManMark = Class("Agent.Defender.ManMark", Base)

local debug = require "../base/debug"
local Field = require "../base/field"
local Referee = require "../base/referee"
local World = require "../base/world"
local vis = require "../base/vis"
local Goal = require "observer/goal"
local CenterBack = require "task/defender/centerback"
local Duel = require "task/shared/duel"
local ManMarkTask = require "task/defender/manmark"
local Defense = require "util/defense"


function ManMark:_stop()
	self._opp = nil
	self._restartTask = true
	self._wasCenterback = false
end

function ManMark:check()
	local role = self._inbox.roleAssignment().trainer
	return role and role.name == "ManMark"
end

function ManMark:_updateTask()
	local newOpp = self._inbox.roleAssignment().trainer.params[1]
	self._restartTask = newOpp ~= self._opp
	self._opp = newOpp
	local wasCenterback = self._wasCenterback
	self._wasCenterback = false

	debug.set("target", self._opp.id)
	local dest = Defense.manMarkPos(self._opp)

	-- try to intercept a possible goal shot
	local _, _, _, passReceivers = Goal.predictShot()
	local passReceiver = passReceivers[1] and passReceivers[1].robot
	if Defense.dangerousBallTowardsDefense() or self._opp == passReceiver then
		local defenseAreaIntersection = Field.intersectRayDefenseArea(World.Ball.pos,
			World.Ball.pos + World.Ball.speed, 0, true)
		if defenseAreaIntersection and World.Ball.pos:distanceTo(defenseAreaIntersection)
			> World.Ball.pos:distanceTo(self._robot.pos)
			and (self._robot.pos - World.Ball.pos):dot(World.Ball.speed) > 0 then
			return Duel
		end
	end

	local color = World.TeamIsBlue and vis.colors.blueHalf or vis.colors.yellowHalf
	vis.addCircle("a/d/manmark: Target", dest, 0.1, color)
	vis.addPath("a/d/manmark: Target", {self._robot.pos, dest, self._opp.pos}, color)

	-- use centerback positioning if the destination pos would be too close to our defense area
	local markingPosDefenseDist = Field.distanceToFriendlyDefenseArea(dest, self._opp.radius)
	local markingPosNearLow = CenterBack.distanceToDefenseArea() + Defense.MARKING_DISTANCE
	local markingPosNearHigh = markingPosNearLow + 2 * self._robot.radius
	local markingPosThreshold = wasCenterback and markingPosNearHigh or markingPosNearLow
	local oppDefenseDist = Field.distanceToFriendlyDefenseArea(self._opp.pos, self._opp.radius)
	if markingPosDefenseDist < markingPosThreshold or oppDefenseDist <= 0 or Referee.isStopState() or Referee.isFriendlyFreeKickState()
			or World.RefereeState == "KickoffOffensivePrepare" or World.RefereeState == "KickoffOffensive" then
		self._wasCenterback = true
		return CenterBack, { self._opp }, self._restartTask
	end

	-- if we are still near the defense area but want to move away, disguise as a centerback
	local selfDefenseDist = Field.distanceToFriendlyDefenseArea(self._robot.pos, self._robot.radius)
	if selfDefenseDist < CenterBack.distanceToDefenseArea() + self._robot.radius + 0.03 then
		local groupApplication = { name = "centerback", payload = nil }
		self._send.groupApplication("trainer", groupApplication)
	end

	return ManMarkTask, { self._opp }, self._restartTask
end

return ManMark

File: strategy/typescript/glados/agent/defender/manmark.ts
import * as debug from "base/debug";
import {Robot} from "base/robot";
import {Position, Vector} from "base/vector";
import * as Field from "base/field";
import * as Referee from "base/referee";
import * as World from "base/world";
import * as vis from "base/vis";
import {Behavior} from "glados/agent/base/behavior";
import {MessageType} from "glados/control/messaging";
import * as Goal from "glados/observer/goal";
import {CenterBack} from "glados/task/defender/centerback"
import {Duel} from "glados/task/shared/duel";
import {ManMark as ManMarkTask} from "glados/task/defender/manmark";
import {Task} from "glados/task/base";
import * as Defense from "glados/util/defense";


export class ManMark extends Behavior {
	_opp: Robot | undefined = undefined;
	_restartTask: boolean = true;
	_wasCenterback: boolean = false;
	_manmarkInfo: {pos: Position, id: number} = {pos: new Vector(0, 0), id: 0};

	_stop () {
		this._opp = undefined;
		this._restartTask = true;
		this._wasCenterback = false;
		this._manmarkInfo = {pos: new Vector(0, 0), id: 0};
	}

	check (): boolean {
		let role = this._messaging.receiveTrainer(MessageType.roleAssignment);
		return role != undefined && role.name === "ManMark";
	}

	_updateTask (): [typeof Task] {
		let roleAssignment = this._messaging.receiveTrainer(MessageType.roleAssignment);
		if (roleAssignment == undefined || roleAssignment.name !== "ManMark") {
			throw new Error();
		}
		let newOpp = roleAssignment.params[0];
		this._restartTask = newOpp != this._opp;
		this._opp = newOpp;
		let wasCenterback = this._wasCenterback;
		this._wasCenterback = false;

		debug.set("target", this._opp.id);
		let dest = Defense.manMarkPos(this._opp);

		// try to intercept a possible goal shot if we are no centerback
		let isCB = this._messaging.receiveTrainer(MessageType.centerBackPosTarget);
		if (isCB == undefined) {
			let passReceivers = Goal.predictShot()[3];
			let passReceiver = passReceivers != undefined && passReceivers.length > 0 ? passReceivers[0].robot : undefined;
			if (Defense.dangerousBallTowardsDefense() || this._opp == passReceiver) {
				let defenseAreaIntersection = Field.intersectRayDefenseArea(World.Ball.pos, World.Ball.speed, 0, true)[0];
				if (defenseAreaIntersection && World.Ball.pos.distanceTo(defenseAreaIntersection)
						> World.Ball.pos.distanceTo(this._robot.pos)
						&&  (this._robot.pos - World.Ball.pos).dot(World.Ball.speed) > 0) {
					return [Duel];
				}
			}
		}

		let color = World.TeamIsBlue ? vis.colors.blueHalf : vis.colors.yellowHalf
		vis.addCircle("a/d/manmark: Target", dest, 0.1, color);
		vis.addPath("a/d/manmark: Target", [this._robot.pos, dest, this._opp.pos], color);

		// use centerback positioning if the destination pos would be too close to our defense area
		let markingPosDefenseDist = Field.distanceToFriendlyDefenseArea(dest, this._opp.radius);
		let markingPosNearLow = Defense.centerBackDistanceToDefenseArea() + Defense.MARKING_DISTANCE;
		let markingPosNearHigh = markingPosNearLow + 2 * this._robot.radius;
		let markingPosThreshold = wasCenterback ? markingPosNearHigh : markingPosNearLow;
		let oppDefenseDist = Field.distanceToFriendlyDefenseArea(this._opp.pos, this._opp.radius);
		if (markingPosDefenseDist < markingPosThreshold || oppDefenseDist <= 0 || Referee.isStopState() || Referee.isFriendlyFreeKickState()
				 ||  World.RefereeState === "KickoffOffensivePrepare" || World.RefereeState === "KickoffOffensive") {
			this._wasCenterback = true;
			// for interpreting debug outputs
			this._manmarkInfo.id = this._opp.id;
			this._manmarkInfo.pos = dest;
			return [CenterBack, [ this._manmarkInfo ], this._restartTask];
		}

		// if we are still near the defense area but want to move away, disguise as a centerback
		let selfDefenseDist = Field.distanceToFriendlyDefenseArea(this._robot.pos, this._robot.radius);
		if (selfDefenseDist < Defense.centerBackDistanceToDefenseArea() + this._robot.radius + 0.03) {
			// TODO EVACUATE
			this._messaging.sendToTrainerRepeated(MessageType.groupApplication, { name: "centerback", payload: undefined });
		}

		return [ManMarkTask, [ this._opp ], this._restartTask];
	}
}